### PR TITLE
fix a client-go bug which could casue kubectl panic

### LINKED ISSUE #72941: panic in kubectl get ... -o custom-columns
**What happened**:
While using output format custom columns, I used array index -1 which caused the following panic:

```
panic: reflect.Value.Slice: slice index out of bounds

goroutine 1 [running]:
reflect.Value.Slice(0x1465900, 0xc4205b0da0, 0x97, 0x4, 0x0, 0x97, 0x240e900, 0x0)
	/usr/lib64/go/src/reflect/value.go:1597 +0x28d
k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath.(*JSONPath).evalArray(0xc420492770, 0xc4205b1280, 0x1, 0x1, 0xc420964780, 0xc4205b1280, 0x1, 0x1, 0x0, 0x0)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath/jsonpath.go:273 +0x43e
k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath.(*JSONPath).walk(0xc420492770, 0xc4205b1280, 0x1, 0x1, 0x189e120, 0xc420964780, 0xc4205b1280, 0x1, 0x1, 0x0, ...)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath/jsonpath.go:148 +0x79f
k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath.(*JSONPath).evalList(0xc420492770, 0xc4205b1220, 0x1, 0x1, 0xc4205b1120, 0xc42085b300, 0x0, 0x0, 0x0, 0xc42085b318)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath/jsonpath.go:202 +0xbb
k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath.(*JSONPath).walk(0xc420492770, 0xc4205b1220, 0x1, 0x1, 0x189e2e0, 0xc4205b1120, 0x0, 0x0, 0xc42085b408, 0x40fc89, ...)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath/jsonpath.go:142 +0x8ae
k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath.(*JSONPath).FindResults(0xc420492770, 0x1541ae0, 0xc4207ca150, 0x18b8420, 0xc420796010, 0x1, 0x10000000046b16c, 0xc42085b4b0)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/client-go/util/jsonpath/jsonpath.go:89 +0x1d4
k8s.io/kubernetes/pkg/printers.(*CustomColumnsPrinter).printOneObject(0xc4200e5140, 0x189d220, 0xc420796010, 0xc420796038, 0x1, 0x1, 0x1893a00, 0xc42074e600, 0xc42085b708, 0x1269b7d)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/printers/customcolumn.go:221 +0x585
k8s.io/kubernetes/pkg/printers.(*CustomColumnsPrinter).PrintObj(0xc4200e5140, 0x189d220, 0xc420796010, 0x1893a00, 0xc42074e600, 0x0, 0x0)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/printers/customcolumn.go:196 +0x740
k8s.io/kubernetes/pkg/printers.(ResourcePrinter).PrintObj-fm(0x189d220, 0xc420796010, 0x18937e0, 0xc4200c0008, 0xc420796010, 0xc420016701)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:252 +0x57
k8s.io/kubernetes/pkg/printers.ResourcePrinterFunc.PrintObj(0xc4205b10a0, 0x189d220, 0xc420796010, 0x18937e0, 0xc4200c0008, 0x0, 0xc4208a77e8)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/printers/interface.go:38 +0x4e
k8s.io/kubernetes/pkg/kubectl/cmd/get.(*GetOptions).printGeneric(0xc4208dc1c0, 0xc420621680, 0x1, 0x1)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:725 +0xc26
k8s.io/kubernetes/pkg/kubectl/cmd/get.(*GetOptions).Run(0xc4208dc1c0, 0x18c6dc0, 0xc420547560, 0xc4208fb180, 0xc4203ef020, 0x2, 0x6, 0x0, 0x0)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:343 +0x13cd
k8s.io/kubernetes/pkg/kubectl/cmd/get.NewCmdGet.func1(0xc4208fb180, 0xc4203ef020, 0x2, 0x6)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/pkg/kubectl/cmd/get/get.go:159 +0x115
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).execute(0xc4208fb180, 0xc4203eeae0, 0x6, 0x6, 0xc4208fb180, 0xc4203eeae0)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:760 +0x2c1
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc4200fef00, 0xc4205f4060, 0x12a05f200, 0xc4208a7ee8)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:846 +0x30a
k8s.io/kubernetes/vendor/github.com/spf13/cobra.(*Command).Execute(0xc4200fef00, 0x17ea8b0, 0x23f0060)
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/spf13/cobra/command.go:794 +0x2b
main.main()
	/builddir/build/BUILD/kubectl-1.11.1-1/_output/local/go/src/k8s.io/kubernetes/cmd/kubectl/kubectl.go:50 +0x196
```

**What you expected to happen**:
It should return the last element of the array, just like index -2 returns second to last element from the array.

**How to reproduce it (as minimally and precisely as possible)**:
`$ kubectl get pod <POD_NAME> -o custom-columns=CONTAINER:.spec.containers[-1].name`

**Environment**:
- Kubernetes version: v1.11.6
- Kubectl version: v1.11.1

### PR DESCRIPTION
**Pull request by WanLinghao**
/kind bug

**What this PR does / why we need it**:
   When user try execute command like：
`kubectl get pod test -o custom-columns=CONTAINER:.spec.containers[-1].name`
    It will throw a panic about slice index out of bounds. This patch fix it.

Fixes #72941 

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go [hunk 1]
```diff
@@ -255,7 +255,7 @@ func (j *JSONPath) evalArray(input []reflect.Value, node *ArrayNode) ([]reflect.
 			params[1].Value = value.Len()
 		}
 
-		if params[1].Value < 0 {
+		if params[1].Value <= 0 {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: a few questions:
* how does changing the condition to include `0` relate to a negative index?
* does the same issue exist on line 251?
* we need a test demonstrating the bug without this change, and demonstrating this fixes it

Comment (WanLinghao) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: @liggitt  

- The data stores in params indicate a slice of array in the form of start-end-step. https://github.com/kubernetes/kubernetes/blob/d857790d360c2f9f86a47964017c08e409312551/staging/src/k8s.io/client-go/util/jsonpath/node.go#L140

- Here what we trying to do is dealing with negative index of array, container[-1] means the last container, container[-2] means the second to last....

- For example, when user want to refer container[-2], the data in params would be [-2, -1, 0]. The condition in L251 reverts param[0] to len-2. The condition in L258 reverts param[0] to len-1. So it becomes [len-2, len-1, 0], which equals container[len-2]

- When it comes to container [-1]. the data in params would be [-1, 0, 0]. You can see that the current condition would not revert param[1] which cause the bug. So we should only change the condition in L251.

Comment (WanLinghao) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go]: I will add a test to confirm this

**Comment by yliaog** [comment]
/lgtm

i assume the test would fail without the fix

**Comment by liggitt** [comment]
/hold

this changes behavior for `foo[0:0]`

File: staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go [hunk 2]
```diff
@@ -369,3 +369,76 @@ func TestFilterPartialMatchesSometimesMissingAnnotations(t *testing.T) {
...
+						"name": "fake3"
+					}]}}`)
+
+	var data interface{}
+	err := json.Unmarshal(input, &data)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testJSONPath(
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go]: thanks for these... some additional cases are needed around range behavior (which the current fix changes):

```
	testJSONPath(
		[]jsonpathTest{
			{
				"test containers[0], it equals containers[0]",
				`{.spec.containers[0].name}`,
				data,
				"fake0",
				false,
			},
			{
				"test containers[0:0], it equals the empty set",
				`{.spec.containers[0:0].name}`,
				data,
				"",
				false,
			},
			{
				"test containers[0:-1], it equals containers[0:3]",
				`{.spec.containers[0:-1].name}`,
				data,
				"fake0 fake1 fake2",
				false,
			},
			{
				"test containers[-1:0], it equals containers[3:0]",
				`{.spec.containers[-1:0].name}`,
				data,
				"fake0 fake1 fake2",
				false,
			},
			{
				"test containers[-1], it equals containers[3]",
				`{.spec.containers[-1].name}`,
				data,
				"fake3",
				false,
			},
			{
				"test containers[-1:], it equals containers[3:]",
				`{.spec.containers[-1:].name}`,
				data,
				"fake3",
				false,
			},
			{
				"test containers[-2], it equals containers[2]",
				`{.spec.containers[-2].name}`,
				data,
				"fake2",
				false,
			},
			{
				"test containers[-2:], it equals containers[2:]",
				`{.spec.containers[-2:].name}`,
				data,
				"fake2 fake3",
				false,
			},
			{
				"test containers[-3], it equals containers[1]",
				`{.spec.containers[-3].name}`,
				data,
				"fake1",
				false,
			},
			{
				"test containers[-4], it equals containers[0]",
				`{.spec.containers[-4].name}`,
				data,
				"fake0",
				false,
			},
			{
				"test containers[-4:], it equals containers[0:]",
				`{.spec.containers[-4:].name}`,
				data,
				"fake0 fake1 fake2 fake3",
				false,
			},
			{
				"test containers[-5], expect a error cause it out of bounds",
				`{.spec.containers[-5].name}`,
				data,
				"",
				true, // expect error
			},
		},
		false,
		t,
	)
```

Comment (liggitt) [reply staging/src/k8s.io/client-go/util/jsonpath/jsonpath_test.go]: also, catching panics in the test makes failures easier to see:
```
func testJSONPath(tests []jsonpathTest, allowMissingKeys bool, t *testing.T) {
	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			defer func() {
				if err := recover(); err != nil {
					t.Error(err)
				}
			}()
			j := New(test.name)
			j.AllowMissingKeys(allowMissingKeys)
			err := j.Parse(test.template)
			if err != nil {
				t.Errorf("in %s, parse %s error %v", test.name, test.template, err)
			}
			buf := new(bytes.Buffer)
			err = j.Execute(buf, test.input)
			if test.expectError {
				if test.expectError && err == nil {
					t.Errorf("in %s, expected execute error", test.name)
				}
				return
			} else if err != nil {
				t.Errorf("in %s, execute error %v", test.name, err)
			}
			out := buf.String()
			if out != test.expect {
				t.Errorf(`in %s, expect to get "%s", got "%s"`, test.name, test.expect, out)
			}
		})
	}
}
```

**Comment by liggitt** [comment]
I think the increment on 0 needs to only happen in the `foo[-1]` case. We can do this by adding a `Derived` attribute to the `ParamsEntry` type, and setting it in parser.go here:

```diff
diff --git a/staging/src/k8s.io/client-go/util/jsonpath/parser.go b/staging/src/k8s.io/client-go/util/jsonpath/parser.go
index 99b45849c9..93d58295ff 100644
--- a/staging/src/k8s.io/client-go/util/jsonpath/parser.go
+++ b/staging/src/k8s.io/client-go/util/jsonpath/parser.go
@@ -325,6 +325,7 @@ Loop:
                        if i == 1 {
                                params[i].Known = true
                                params[i].Value = params[0].Value + 1
+                               params[i].Derived = true
```

then check it like this (this also fixes the panic if start > end, or if start==end and is out of bounds):
```diff
diff --git a/staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go b/staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go
index d4d9737414..ed5a45aa73 100644
--- a/staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go
+++ b/staging/src/k8s.io/client-go/util/jsonpath/jsonpath.go
@@ -254,8 +254,7 @@ func (j *JSONPath) evalArray(input []reflect.Value, node *ArrayNode) ([]reflect.
                if !params[1].Known {
                        params[1].Value = value.Len()
                }
-
-               if params[1].Value < 0 {
+               if params[1].Value < 0 || (params[1].Value == 0 && params[1].Derived) {
                        params[1].Value += value.Len()
                }

@@ -267,6 +266,11 @@ func (j *JSONPath) evalArray(input []reflect.Value, node *ArrayNode) ([]reflect.
                        if params[1].Value > sliceLength || params[1].Value < 0 {
                                return input, fmt.Errorf("array index out of bounds: index %d, length %d", params[1].Value-1, sliceLength)
                        }
+                       if params[0].Value > params[1].Value {
+                               return input, fmt.Errorf("starting index %d is greater than ending index %d", params[1].Value, params[1].Value)
+                       }
+               } else {
+                       return result, nil
                }

                if !params[2].Known {
```

**Comment by WanLinghao** [comment]
@liggitt  Thanks for your help! All comments addressed, PTAL

**Pull request merged**